File: backend/src/main/kotlin/com/task/presentation/TaskExecutions.kt
package com.task.presentation

import com.task.domain.member.MemberId
import com.task.domain.taskExecution.TaskExecutionId
import com.task.usecase.taskExecution.assign.AssignTaskExecutionUseCase
import com.task.usecase.taskExecution.cancel.CancelTaskExecutionUseCase
import com.task.usecase.taskExecution.complete.CompleteTaskExecutionUseCase
import com.task.usecase.taskExecution.get.GetTaskExecutionUseCase
import com.task.usecase.taskExecution.get.GetTaskExecutionsUseCase
import com.task.usecase.taskExecution.start.StartTaskExecutionUseCase
import io.ktor.http.HttpStatusCode
import io.ktor.resources.Resource
import io.ktor.server.application.call
import io.ktor.server.request.receive
import io.ktor.server.response.respond
import io.ktor.server.routing.Route
import io.ktor.server.resources.get
import io.ktor.server.resources.post
import kotlinx.serialization.Serializable
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.UUID

@Resource("api/task-executions")
class TaskExecutions {

    @Resource("")
    class List(
        val parent: TaskExecutions = TaskExecutions(),
        val limit: Int = 20,
        val offset: Int = 0,
        val scheduledDate: String? = null,  // YYYY-MM-DD形式
        val status: String? = null,         // NOT_STARTED, IN_PROGRESS, COMPLETED, CANCELLED
        val assigneeMemberId: String? = null // UUID形式
    )

    @Resource("/{taskExecutionId}")
    class Get(
        val parent: TaskExecutions = TaskExecutions(),
        val taskExecutionId: String
    )

    @Resource("/{taskExecutionId}/start")
    class Start(
        val parent: TaskExecutions = TaskExecutions(),
        val taskExecutionId: String
    ) {
        @Serializable
        data class Request(
            val memberId: String
        )
    }

    @Resource("/{taskExecutionId}/complete")
    class Complete(
        val parent: TaskExecutions = TaskExecutions(),
        val taskExecutionId: String
    ) {
        @Serializable
        data class Request(
            val memberId: String
        )
    }

    @Resource("/{taskExecutionId}/cancel")
    class Cancel(
        val parent: TaskExecutions = TaskExecutions(),
        val taskExecutionId: String
    )

    @Resource("/{taskExecutionId}/assign")
    class Assign(
        val parent: TaskExecutions = TaskExecutions(),
        val taskExecutionId: String
    ) {
        @Serializable
        data class Request(
            val memberId: String
        )
    }
}

@Serializable
data class TaskExecutionDto(
    val id: String,
    val taskDefinitionId: String,
    val assigneeMemberId: String?,
    val scheduledDate: String,
    val status: String,
    val taskSnapshot: TaskSnapshotDto?,
    val startedAt: String?,
    val completedAt: String?,
    val completedByMemberId: String?
)

@Serializable
data class TaskSnapshotDto(
    val name: String,
    val description: String?,
    val scheduledStartTime: String,
    val scheduledEndTime: String,
    val definitionVersion: Int,
    val capturedAt: String
)

@Serializable
data class GetTaskExecutionsResponse(
    val taskExecutions: kotlin.collections.List<TaskExecutionDto>,
    val total: Int,
    val hasMore: Boolean
)

private val dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd").withZone(ZoneId.systemDefault())
private val instantFormatter = DateTimeFormatter.ISO_INSTANT

fun Route.taskExecutions() {

    get<TaskExecutions.List> { resource ->
        // クエリパラメータからFilterSpecを構築
        val filter = GetTaskExecutionsUseCase.FilterSpec(
            scheduledDate = resource.scheduledDate?.let { LocalDate.parse(it) },
            status = resource.status,
            assigneeMemberId = resource.assigneeMemberId?.let { MemberId(UUID.fromString(it)) }
        )

        val output = instance<GetTaskExecutionsUseCase>().execute(
            GetTaskExecutionsUseCase.Input(
                limit = resource.limit,
                offset = resource.offset,
                filter = filter
            )
        )

        val taskExecutions = output.items.map { item ->
            TaskExecutionDto(
                id = item.id.value.toString(),
                taskDefinitionId = item.taskDefinitionId.value.toString(),
                assigneeMemberId = item.assigneeMemberId?.value?.toString(),
                scheduledDate = dateFormatter.format(item.scheduledDate),
                status = item.status,
                taskSnapshot = item.snapshot?.let { snapshot ->
                    TaskSnapshotDto(
                        name = snapshot.name,
                        description = snapshot.description,
                        scheduledStartTime = instantFormatter.format(snapshot.scheduledStartTime),
                        scheduledEndTime = instantFormatter.format(snapshot.scheduledEndTime),
                        definitionVersion = snapshot.definitionVersion,
                        capturedAt = instantFormatter.format(snapshot.capturedAt)
                    )
                },
                startedAt = item.startedAt?.let { instantFormatter.format(it) },
                completedAt = item.completedAt?.let { instantFormatter.format(it) },
                completedByMemberId = item.completedByMemberId?.value?.toString()
            )
        }

        call.respond(
            HttpStatusCode.OK,
            GetTaskExecutionsResponse(
                taskExecutions = taskExecutions,
                total = output.totalCount,
                hasMore = output.totalCount > resource.offset + resource.limit
            )
        )
    }

    get<TaskExecutions.Get> { resource ->
        val output = instance<GetTaskExecutionUseCase>().execute(
            GetTaskExecutionUseCase.Input(
                id = TaskExecutionId.from(resource.taskExecutionId)
            )
        )

        if (output == null) {
            call.respond(HttpStatusCode.NotFound, mapOf("error" to "TaskExecution not found"))
            return@get
        }

        call.respond(
            HttpStatusCode.OK,
            TaskExecutionDto(
                id = output.id.value.toString(),
                taskDefinitionId = output.taskDefinitionId.value.toString(),
                assigneeMemberId = output.assigneeMemberId?.value?.toString(),
                scheduledDate = dateFormatter.format(output.scheduledDate),
                status = output.status,
                taskSnapshot = output.snapshot?.let { snapshot ->
                    TaskSnapshotDto(
                        name = snapshot.name,
                        description = snapshot.description,
                        scheduledStartTime = instantFormatter.format(snapshot.scheduledStartTime),
                        scheduledEndTime = instantFormatter.format(snapshot.scheduledEndTime),
                        definitionVersion = snapshot.definitionVersion,
                        capturedAt = instantFormatter.format(snapshot.capturedAt)
                    )
                },
                startedAt = output.startedAt?.let { instantFormatter.format(it) },
                completedAt = output.completedAt?.let { instantFormatter.format(it) },
                completedByMemberId = output.completedByMemberId?.value?.toString()
            )
        )
    }

    post<TaskExecutions.Start> { resource ->
        val request = call.receive<TaskExecutions.Start.Request>()

        val output = instance<StartTaskExecutionUseCase>().execute(
            StartTaskExecutionUseCase.Input(
                id = TaskExecutionId.from(resource.taskExecutionId),
                assigneeMemberId = MemberId(UUID.fromString(request.memberId))
            )
        )

        call.respond(
            HttpStatusCode.OK,
            TaskExecutionDto(
                id = output.id.value.toString(),
                taskDefinitionId = output.taskDefinitionId.value.toString(),
                assigneeMemberId = output.assigneeMemberId.value.toString(),
                scheduledDate = dateFormatter.format(output.scheduledDate),
                status = "IN_PROGRESS",
                taskSnapshot = TaskSnapshotDto(
                    name = output.taskSnapshot.frozenName.value,
                    description = output.taskSnapshot.frozenDescription.value,
                    scheduledStartTime = instantFormatter.format(output.taskSnapshot.frozenScheduledTimeRange.startTime),
                    scheduledEndTime = instantFormatter.format(output.taskSnapshot.frozenScheduledTimeRange.endTime),
                    definitionVersion = output.taskSnapshot.definitionVersion,
                    capturedAt = instantFormatter.format(output.taskSnapshot.capturedAt)
                ),
                startedAt = instantFormatter.format(output.startedAt),
                completedAt = null,
                completedByMemberId = null
            )
        )
    }

    post<TaskExecutions.Complete> { resource ->
        val request = call.receive<TaskExecutions.Complete.Request>()

        val output = instance<CompleteTaskExecutionUseCase>().execute(
            CompleteTaskExecutionUseCase.Input(
                id = TaskExecutionId.from(resource.taskExecutionId),
                completedMemberId = MemberId(UUID.fromString(request.memberId))
            )
        )

        call.respond(
            HttpStatusCode.OK,
            TaskExecutionDto(
                id = output.id.value.toString(),
                taskDefinitionId = output.taskDefinitionId.value.toString(),
                assigneeMemberId = output.assigneeMemberId.value.toString(),
                scheduledDate = dateFormatter.format(output.scheduledDate),
                status = "COMPLETED",
                taskSnapshot = TaskSnapshotDto(
                    name = output.taskSnapshot.frozenName.value,
                    description = output.taskSnapshot.frozenDescription.value,
                    scheduledStartTime = instantFormatter.format(output.taskSnapshot.frozenScheduledTimeRange.startTime),
                    scheduledEndTime = instantFormatter.format(output.taskSnapshot.frozenScheduledTimeRange.endTime),
                    definitionVersion = output.taskSnapshot.definitionVersion,
                    capturedAt = instantFormatter.format(output.taskSnapshot.capturedAt)
                ),
                startedAt = instantFormatter.format(output.startedAt),
                completedAt = instantFormatter.format(output.completedAt),
                completedByMemberId = output.completedMemberId.value.toString()
            )
        )
    }

    post<TaskExecutions.Cancel> { resource ->
        val output = instance<CancelTaskExecutionUseCase>().execute(
            CancelTaskExecutionUseCase.Input(
                id = TaskExecutionId.from(resource.taskExecutionId)
            )
        )

        call.respond(
            HttpStatusCode.OK,
            mapOf(
                "id" to output.id.value.toString(),
                "taskDefinitionId" to output.taskDefinitionId.value.toString(),
                "scheduledDate" to dateFormatter.format(output.scheduleDate),
                "status" to "CANCELLED"
            )
        )
    }

    post<TaskExecutions.Assign> { resource ->
        val request = call.receive<TaskExecutions.Assign.Request>()

        val output = instance<AssignTaskExecutionUseCase>().execute(
            AssignTaskExecutionUseCase.Input(
                id = TaskExecutionId.from(resource.taskExecutionId),
                newAssigneeMemberId = MemberId(UUID.fromString(request.memberId))
            )
        )

        call.respond(
            HttpStatusCode.OK,
            TaskExecutionDto(
                id = output.id.value.toString(),
                taskDefinitionId = output.taskDefinitionId.value.toString(),
                assigneeMemberId = output.assigneeMemberId?.value?.toString(),
                scheduledDate = dateFormatter.format(output.scheduledDate),
                status = output.status,
                taskSnapshot = output.snapshot?.let { snapshot ->
                    TaskSnapshotDto(
                        name = snapshot.name,
                        description = snapshot.description,
                        scheduledStartTime = instantFormatter.format(snapshot.scheduledStartTime),
                        scheduledEndTime = instantFormatter.format(snapshot.scheduledEndTime),
                        definitionVersion = snapshot.definitionVersion,
                        capturedAt = instantFormatter.format(snapshot.capturedAt)
                    )
                },
                startedAt = output.startedAt?.let { instantFormatter.format(it) },
                completedAt = output.completedAt?.let { instantFormatter.format(it) },
                completedByMemberId = output.completedByMemberId?.value?.toString()
            )
        )
    }
}

File: frontend/src/mocks/taskExecutions.ts
/**
 * タスク実行関連のモックデータ
 *
 * @note バックエンドのTaskExecution APIが実装されるまでの暫定データ
 * @see doc/task-execution-api.md
 */

import type { TaskExecution, ExecutionStatus } from '../types'
import { toISODateString } from '../utils'

/**
 * 今日の日付を取得（モックデータ用）
 */
const TODAY = toISODateString(new Date())

/**
 * モックタスク実行データ
 *
 * Dashboard画面で使用される「今日のタスク」のサンプルデータ
 */
export const MOCK_TASK_EXECUTIONS: TaskExecution[] = [
  {
    id: 'exec-1',
    taskDefinitionId: 'def-1',
    assigneeMemberId: '2',
    scheduledDate: TODAY,
    status: 'COMPLETED',
    taskSnapshot: {
      name: 'お風呂掃除',
      description: '浴槽と床を洗う',
      scheduledStartTime: '2025-01-01T20:00:00Z',
      scheduledEndTime: '2025-01-01T20:15:00Z',
      definitionVersion: 1,
      capturedAt: new Date().toISOString(),
    },
    startedAt: new Date(Date.now() - 3600000).toISOString(), // 1時間前
    completedAt: new Date(Date.now() - 2700000).toISOString(), // 45分前
    completedByMemberId: '2',
    createdAt: new Date(Date.now() - 86400000).toISOString(), // 1日前
    updatedAt: new Date(Date.now() - 2700000).toISOString(),
  },
  {
    id: 'exec-2',
    taskDefinitionId: 'def-2',
    assigneeMemberId: '3',
    scheduledDate: TODAY,
    status: 'IN_PROGRESS',
    taskSnapshot: {
      name: '洗濯物を干す',
      description: '洗濯機を回してからベランダに干す',
      scheduledStartTime: '2025-01-01T08:00:00Z',
      scheduledEndTime: '2025-01-01T08:20:00Z',
      definitionVersion: 1,
      capturedAt: new Date().toISOString(),
    },
    startedAt: new Date(Date.now() - 1800000).toISOString(), // 30分前
    createdAt: new Date(Date.now() - 86400000).toISOString(),
    updatedAt: new Date(Date.now() - 1800000).toISOString(),
  },
  {
    id: 'exec-3',
    taskDefinitionId: 'def-3',
    scheduledDate: TODAY,
    status: 'NOT_STARTED',
    taskSnapshot: {
      name: '夕食の準備',
      description: 'カレーを作る',
      scheduledStartTime: '2025-01-01T17:30:00Z',
      scheduledEndTime: '2025-01-01T18:15:00Z',
      definitionVersion: 1,
      capturedAt: '',
    },
    createdAt: new Date(Date.now() - 86400000).toISOString(),
    updatedAt: new Date(Date.now() - 86400000).toISOString(),
  },
  {
    id: 'exec-4',
    taskDefinitionId: 'def-4',
    assigneeMemberId: '2',
    scheduledDate: TODAY,
    status: 'NOT_STARTED',
    taskSnapshot: {
      name: 'ゴミ出し',
      description: '燃えるゴミを出す',
      scheduledStartTime: '2025-01-01T07:45:00Z',
      scheduledEndTime: '2025-01-01T07:50:00Z',
      definitionVersion: 1,
      capturedAt: '',
    },
    createdAt: new Date(Date.now() - 86400000).toISOString(),
    updatedAt: new Date(Date.now() - 86400000).toISOString(),
  },
  {
    id: 'exec-5',
    taskDefinitionId: 'def-5',
    assigneeMemberId: '4',
    scheduledDate: TODAY,
    status: 'CANCELLED',
    taskSnapshot: {
      name: '買い物',
      description: '牛乳と卵を買う',
      scheduledStartTime: '2025-01-01T16:00:00Z',
      scheduledEndTime: '2025-01-01T17:00:00Z',
      definitionVersion: 1,
      capturedAt: new Date().toISOString(),
    },
    startedAt: new Date(Date.now() - 7200000).toISOString(), // 2時間前
    createdAt: new Date(Date.now() - 86400000).toISOString(),
    updatedAt: new Date(Date.now() - 3600000).toISOString(),
  },
]

/**
 * タスク実行IDからタスク実行を取得
 */
export function getMockTaskExecutionById(id: string): TaskExecution | undefined {
  return MOCK_TASK_EXECUTIONS.find((exec) => exec.id === id)
}

/**
 * 指定日のタスク実行を取得
 */
export function getMockTaskExecutionsByDate(date: string): TaskExecution[] {
  return MOCK_TASK_EXECUTIONS.filter((exec) => exec.scheduledDate === date)
}

/**
 * 指定ステータスのタスク実行を取得
 */
export function getMockTaskExecutionsByStatus(status: ExecutionStatus): TaskExecution[] {
  return MOCK_TASK_EXECUTIONS.filter((exec) => exec.status === status)
}

/**
 * 指定メンバーのタスク実行を取得
 */
export function getMockTaskExecutionsByMember(memberId: string): TaskExecution[] {
  return MOCK_TASK_EXECUTIONS.filter((exec) => exec.assigneeMemberId === memberId)
}

/**
 * 今日のタスクサマリーを取得
 */
export function getMockTodayTaskSummary(): {
  total: number
  completed: number
  inProgress: number
  notStarted: number
  cancelled: number
} {
  const todayTasks = getMockTaskExecutionsByDate(TODAY)

  return {
    total: todayTasks.length,
    completed: todayTasks.filter((t) => t.status === 'COMPLETED').length,
    inProgress: todayTasks.filter((t) => t.status === 'IN_PROGRESS').length,
    notStarted: todayTasks.filter((t) => t.status === 'NOT_STARTED').length,
    cancelled: todayTasks.filter((t) => t.status === 'CANCELLED').length,
  }
}